Add Wilight integration with SSDP

leofig-rj opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Breaking change
<!--
  If your PR contains a breaking change for existing users, it is important
  to tell them what breaks, how to make it work again and why we did this.
  This piece of text is published with the release notes, so it helps if you
  write it towards our users, not us.
  Note: Remove this section if this PR is NOT a breaking change.
-->


## Proposed change
<!-- 
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->


## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box! 
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Example entry for `configuration.yaml`:
<!--
  Supplying a configuration snippet, makes it easier for a maintainer to test
  your PR. Furthermore, for new integrations, it gives an impression of how
  the configuration would look like.
  Note: Remove this section if this PR does not have an example entry.
-->

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request:
  home-assistant.io:https://github.com/home-assistant/home-assistant.io/pull/13547
  brands PR: https://github.com/home-assistant/brands/pull/1642

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [x] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [x] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [x] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [x] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [x] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

leofig-rj committed:
Wilight initial with SSDP

frenck commented:
Not stale awaiting our review.

MartinHjelmare commented on homeassistant/components/wilight/config_flow.py:
```diff
@@ -0,0 +1,111 @@
+            return self.async_abort(reason="not_wilight")
+
+        component_ok = False
+        for self_component in self._components:
```
Please use a different variable name that is less easy to confuse with `self._components`.
```suggestion
        for wilight_component in self._components:
```

leofig-rj replied:
Done on lines 84 and 86

MartinHjelmare commented on homeassistant/components/wilight/config_flow.py:
```diff
@@ -0,0 +1,111 @@
+        component_ok = False
+        for self_component in self._components:
+            for component in PLATFORMS:
+                if self_component == component:
```
I think we can use `any` for this check.
```py
component_ok = any(
    wilight_component in allowed_components
    for wilight_component in self._components
)
```

leofig-rj replied:
It really looks much more elegant. Done.

MartinHjelmare commented on homeassistant/components/wilight/config_flow.py:
```diff
@@ -0,0 +1,111 @@
+    def _self_update(self, host, serial_number, model_name):
+        self._host = host
+        self._serial_number = serial_number
+        self._device_id = f"WL{serial_number}"
```
We don't need to add `WL` to the unique_id. The core is aware of the integration domain when comparing unique ids.

leofig-rj replied:
I prefer to keep the WL, because that's how it appears in the integrations.
If it is not an impediment, I would like to maintain.

MartinHjelmare replied:
Use a separate title not the same as the unique_id for the entry.

leofig-rj replied:
I'm thinking of replacing self._device_id with self._tittle and on line 85: await self.async_set_unique_id (self._device_id)
change to:
await self.async_set_unique_id (self._serial_number)

What do you think?

leofig-rj replied:
Done

MartinHjelmare commented on homeassistant/components/wilight/light.py:
```diff
@@ -0,0 +1,223 @@
+            entity = WiLightLightDimmer(api_device, index, item_name)
+        elif item["sub_type"] == LIGHT_COLOR:
+            entity = WiLightLightColor(api_device, index, item_name)
+        else:
+            continue
```
The for loop will continue by default when reaching the end.
```suggestion
```

leofig-rj replied:
But I want to add entities except in the case else...

entities.append(entity)

MartinHjelmare replied:
Right.

MartinHjelmare commented on homeassistant/components/wilight/light.py:
```diff
@@ -0,0 +1,223 @@
+    async_add_entities(entities)
+
+
+async def async_unload_entry(hass: HomeAssistant, entry: ConfigEntry):
+    """Unload a config entry."""
+
+    return True
+
+
```
The platform doesn't have an unload api.
```suggestion
```

MartinHjelmare commented on homeassistant/components/wilight/light.py:
```diff
@@ -0,0 +1,223 @@
+    @property
+    def is_on(self):
+        """Return true if device is on."""
+        if "on" in self._status:
+            self._on = self._status["on"]
+        return self._on
```
Why do we need the instance attribute?
```suggestion
        return self._status.get("on")
```

leofig-rj replied:
At startup, self._status = {}.
self._status is updated with IO and may not contain "on", "brightness", "hue" or "saturation" at the beginning.

That is why:

    def __init __ (self, * args, ** kwargs):
        "" "Initialize the device." ""
        WiLightDevice .__ init __ (self, * args, ** kwargs)
        # Initialize the WiLights color.
        self._on = False
        self._brightness = 0
        self._hue = 0
        self._saturation = 100

    @property
    def brightness (self):
        "" "Return the brightness of this light between 0..255." ""
        if "brightness" in self._status:
            self._brightness = int (self._status ["brightness"])
        return self._brightness

    @property
    def hs_color (self):
        "" "Return the hue and saturation color value [float, float]." ""
        if "hue" in self._status:
            self._hue = wilight_to_hass_hue (int (self._status ["hue"]))
        if "saturation" in self._status:
            self._saturation = wilight_to_hass_saturation (
                int (self._status ["saturation"])
            )
        return [self._hue, self._saturation]

    @property
    def is_on (self):
        "" "Return true if device is on." ""
        if "on" in self._status:
            self._on = self._status ["on"]
        return self._on

MartinHjelmare replied:
Using `dict.get` as in my suggestion handles that case.

leofig-rj replied:
OK, Thanks

MartinHjelmare commented on homeassistant/components/wilight/light.py:
```diff
@@ -0,0 +1,223 @@
+class WiLightLightOnOff(WiLightDevice, LightEntity):
+    """Representation of a WiLights light on-off."""
+
+    def __init__(self, *args, **kwargs):
+        """Initialize the device."""
+        WiLightDevice.__init__(self, *args, **kwargs)
+        # Initialize the WiLights onoff.
+        self._on = False
+
```
```suggestion
```

leofig-rj replied:
Same as above

MartinHjelmare commented on homeassistant/components/wilight/light.py:
```diff
@@ -0,0 +1,223 @@
+class WiLightLightDimmer(WiLightDevice, LightEntity):
+    """Representation of a WiLights light dimmer."""
+
+    def __init__(self, *args, **kwargs):
+        """Initialize the device."""
+        WiLightDevice.__init__(self, *args, **kwargs)
+        # Initialize the WiLights dimmer.
+        self._on = False
+        self._brightness = 0
+
```
```suggestion
```

leofig-rj replied:
Same as above

MartinHjelmare commented on homeassistant/components/wilight/light.py:
```diff
@@ -0,0 +1,223 @@
+    @property
+    def brightness(self):
+        """Return the brightness of this light between 0..255."""
+        if "brightness" in self._status:
+            self._brightness = int(self._status["brightness"])
+        return self._brightness
```
```suggestion
        return int(self._status.get("brightness", 0))
```

leofig-rj replied:
Same as above

MartinHjelmare commented on homeassistant/components/wilight/light.py:
```diff
@@ -0,0 +1,223 @@
+    @property
+    def is_on(self):
+        """Return true if device is on."""
+        if "on" in self._status:
+            self._on = self._status["on"]
+        return self._on
```
```suggestion
        return self._status.get("on")
```

MartinHjelmare commented on homeassistant/components/wilight/light.py:
```diff
@@ -0,0 +1,223 @@
+    @property
+    def supported_features(self):
+        """Flag supported features."""
+        return SUPPORT_BRIGHTNESS + SUPPORT_COLOR
```
```suggestion
        return SUPPORT_BRIGHTNESS | SUPPORT_COLOR
```

leofig-rj replied:
OK

MartinHjelmare commented on homeassistant/components/wilight/parent_device.py:
```diff
@@ -0,0 +1,110 @@
+
+        async def connect(api_device):
+            # Set up connection and hook it into HA for reconnect/shutdown.
+            _LOGGER.info("Initiating WiLight connection to %s", api_device.device_id)
```
```suggestion
            _LOGGER.debug("Initiating connection to %s", api_device.device_id)
```

MartinHjelmare commented on homeassistant/components/wilight/parent_device.py:
```diff
@@ -0,0 +1,110 @@
+        host = self._host
+        hass = self._hass
+
+        api_device = await hass.async_add_executor_job(create_api_device, host, 45995)
```
Why is this done in the executor while the rest of the api calls are awaited? Does the library support asyncio?

leofig-rj replied:
Honestly, this was the best way I did to integrate after several tests.
Yes, the library supports asyncio.

MartinHjelmare replied:
Do we need to make a request to the URL or can we create the device from the info we get from the SSDP discovery in the config flow? We have at least host, serial number and model.

leofig-rj replied:
I was wrong when I replied that the library supports asyncio.
In fact, the part that deals with discovery, until creating the device, does not.
Who supports asyncio is the communication part (API) after creating the device
Hence the need for async_add_executor_job.
Today I can even create a device from host, serial number and model, but I would also need to get model_number from ssdp.
It turns out that I plan to get more device information in the future to import the names of the child devices from my application. Homeassinstant SSDP does not support this.
Therefore, I intend to keep async_add_executor_job if that is not a hindrance to approve the integration.

MartinHjelmare replied:
I recommend using aiohttp for the requests instead of requests. Then the library would be totally async. It's not a requirement.

leofig-rj replied:
How this will imply major changes and tests in the library. I will try, but I would like it to go to a new PR when I add new components.
Can we approve the current one as it is?

MartinHjelmare replied:
Yes, it's ok as is.

MartinHjelmare commented on homeassistant/components/wilight/parent_device.py:
```diff
@@ -0,0 +1,110 @@
+
+def create_api_device(host, port):
+    """Create an API Device."""
+    url = f"http://{host}:{port}/wilight.xml"
```
Why is this string not created in the library? This is protocol level information.

leofig-rj replied:
I could create a new function in the API like:
device = pywilight.device_from_host_port (host, port)

But I would have to do a new revision in the API too.

If this is not a major inconvenience, could it be in a new PullRequest when adding new components?

MartinHjelmare replied:
This PR isn't merged yet. We can wait for the library bump.

leofig-rj replied:
I will modify the library to support this.

MartinHjelmare commented on homeassistant/components/wilight/config_flow.py:
```diff
@@ -0,0 +1,111 @@
+        self._serial_number = serial_number
+        self._device_id = f"WL{serial_number}"
+        self._model_name = model_name
+        self._components = pywilight.get_components_from_model(model_name)
```
Does this do I/O?

leofig-rj replied:
A WiLight device can have more than one component, for example light and fan.
In this first integration I am only including light, taking into account the guidelines for not having a very extensive initial integration.
pywilight.get_components_from_model (model_name)
It is a function that lists the components that that WiLight model has.

MartinHjelmare replied:
So it won't do I/O?

MartinHjelmare replied:
I checked the library. It's safe.

MartinHjelmare commented on tests/components/wilight/test_init.py:
```diff
@@ -0,0 +1,66 @@
+def mock_dummy_create_api_device():
+    """Mock a valid api_devce."""
+
+    device = pywilight.discovery.wilight_from_discovery(
```
Won't this do I/O?

leofig-rj replied:
My API has a simulation mode for testing that is activated with:

    device.set_dummy (True)

Making the patch:
     with patch (
         "homeassistant.components.wilight.parent_device.create_api_device",
         return_value = device,
     ):
         yield device

In this way I simulate the IOs ...

MartinHjelmare replied:
Ok. We should patch our own code, ie `create_api_device`. Patch the library instead to return the dummy device.

leofig-rj replied:
I'll check how I do this.

leofig-rj committed:
Update homeassistant/components/wilight/__init__.py

Co-authored-by: Martin Hjelmare <marhje52@gmail.com>

leofig-rj committed:
Update homeassistant/components/wilight/__init__.py

Co-authored-by: Martin Hjelmare <marhje52@gmail.com>

leofig-rj committed:
Update homeassistant/components/wilight/__init__.py

Co-authored-by: Martin Hjelmare <marhje52@gmail.com>

leofig-rj committed:
Update homeassistant/components/wilight/parent_device.py

Co-authored-by: Martin Hjelmare <marhje52@gmail.com>

leofig-rj commented on homeassistant/components/wilight/__init__.py:
```diff
@@ -0,0 +1,136 @@
+        """Initialize the device."""
+        # WiLight specific attributes for every component type
+        self._device_id = api_device.device_id
+        self._swversion = api_device.swversion
```
Updated on line 77 and 108

leofig-rj commented on homeassistant/components/wilight/config_flow.py:
```diff
@@ -0,0 +1,111 @@
+        self._model_name = model_name
+        self._components = pywilight.get_components_from_model(model_name)
+        text = ""
+        for component in self._components:
+            if text == "":
+                text = component
+            else:
+                text = ", ".join((text, component))
+        self._components_text = text
```
Doing what you suggest will not start with ","?
I don't want to start like this.

MartinHjelmare replied:
It won't start with a comma. Try it in a python interpreter.

MartinHjelmare replied:
It's also easier to strip an offending comma at an end after the join if that would be needed.

leofig-rj replied:
It won't start with a comma. Try it in a python interpreter. -> Right!!!
Done

leofig-rj committed:
Update homeassistant/components/wilight/config_flow.py

Co-authored-by: Martin Hjelmare <marhje52@gmail.com>

leofig-rj commented on homeassistant/components/wilight/__init__.py:
```diff
@@ -0,0 +1,136 @@
+async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry):
+    """Set up a wilight config entry."""
+
+    if CONF_HOST not in entry.data:
```
Do you suggest removing?

     if CONF_HOST not in entry.data:
         return False

MartinHjelmare replied:
Yes.

leofig-rj replied:
Done

leofig-rj committed:
Update homeassistant/components/wilight/light.py

Co-authored-by: Martin Hjelmare <marhje52@gmail.com>

leofig-rj commented on .coveragerc:
```diff
@@ -888,6 +888,7 @@ omit =
     homeassistant/components/wemo/*
     homeassistant/components/whois/sensor.py
     homeassistant/components/wiffi/*
+    homeassistant/components/wilight/*
```
I'm sorry, but I didn't understand what you meant.
Could you please explain otherwise?
Thank you

MartinHjelmare replied:
Remove the star. Only add the modules that we don't want to calculate coverage for. If we cover all modules with tests we don't need to exclude any modules from coverage calculation.

leofig-rj replied:
Removed homeassistant/components/wilight/*

leofig-rj committed:
As Martin suggested

leofig-rj committed:
Update homeassistant/components/wilight/light.py

Co-authored-by: Martin Hjelmare <marhje52@gmail.com>

leofig-rj committed:
Update config_flow.py

leofig-rj committed:
As requested by Martin

leofig-rj commented on homeassistant/components/wilight/config_flow.py:
```diff
@@ -0,0 +1,111 @@
+    VERSION = 1
+    CONNECTION_CLASS = CONN_CLASS_LOCAL_PUSH
+
+    # pylint: disable=no-member # https://github.com/PyCQA/pylint/issues/3167
```
Done

leofig-rj committed:
Update config_flow.py

leofig-rj committed:
Update config_flow.py

MartinHjelmare commented on homeassistant/components/wilight/parent_device.py:
```diff
@@ -0,0 +1,110 @@
+            # Set up connection and hook it into HA for reconnect/shutdown.
+            _LOGGER.debug("Initiating connection to %s", api_device.device_id)
+
+            client = await api_device.config_client(
```
This task can take take forever if the connection can't be established. We should either have a timeout or let the library handle the connection and not hold up the integration setup for this. In the latter case we need to schedule the task with `asyncio.create_task`.

MartinHjelmare replied:
I noticed when looking at the library that it uses `time.sleep` in a reconnect function. I don't see that function being used anywhere, but please make sure that we don't use `time.sleep` in the event loop as it will block it.

leofig-rj replied:
The initial connection has a timeout, but the reconnection did have a sleep.
I'm updating the API and exchanging sleep for:
    self._reconnect_timer = self._loop.call_later (self._reconnect_interval, self.do_reconnect)
The self.do_reconnect procedure has a timeout.

MartinHjelmare replied:
Looks like the `while True:` can be removed since there's a `break` at the end:
https://github.com/leofig-rj/pywilight/blob/71fae849d5464a444bef3bf9671046a8f70332b1/pywilight/wilight_device/protocol.py#L488

leofig-rj replied:
You are correct, I will remove. I was testing it and then I was going to review it.

leofig-rj replied:
Done in Version 0.0.65

MartinHjelmare commented:
Test failure is unrelated.

leofig-rj committed:
Update homeassistant/components/wilight/light.py

Co-authored-by: Martin Hjelmare <marhje52@gmail.com>

leofig-rj committed:
Update homeassistant/components/wilight/light.py

Co-authored-by: Martin Hjelmare <marhje52@gmail.com>

leofig-rj committed:
Update homeassistant/components/wilight/light.py

Co-authored-by: Martin Hjelmare <marhje52@gmail.com>

leofig-rj committed:
Update homeassistant/components/wilight/light.py

Co-authored-by: Martin Hjelmare <marhje52@gmail.com>

leofig-rj committed:
Update homeassistant/components/wilight/light.py

Co-authored-by: Martin Hjelmare <marhje52@gmail.com>

leofig-rj committed:
Update light.py

leofig-rj committed:
Update coverage and __init__.py

leofig-rj committed:
Update config_flow.py

leofig-rj committed:
Update config_flow.py

MartinHjelmare commented on homeassistant/components/wilight/config_flow.py:
```diff
@@ -0,0 +1,104 @@
+        if not self._self_update(host, serial_number, model_name):
+            return self.async_abort(reason="not_wilight")
+
+        allowed_components = PLATFORMS
```
Please create a separate constant for `allowed_components`. We want to keep the home assistant API separate from the Wilight API as either of those may change.

leofig-rj replied:
PLATFORMS is a constant that comes from __INIT__.py and not from the Wilight  API.
It serves to indicate which components are allowed by the current integration.
I used allowed_components for your suggestion.
I used the name PLATFORMS, because it was already in a template created by Homeassistant.
Would you like me to rename PLATFORMS to ALLOWED_COMPONENTS in __INIT__.py?

leofig-rj replied:
This is where I see if the components available in the WiLight API are allowed by the current integration.

MartinHjelmare replied:
The `PLATFORMS` constant contains strings for the home assistant api. `allowed_components` should contain strings for the wilight api. They must be created in different separate variables. We have to create a new constant, named eg `ALLOWED_WILIGHT_COMPONENTS`. Don't change `PLATFORMS`.

leofig-rj replied:
Honestly I don't see it like that, I see a check of the components of the WiLight API (self._components) against the components of the Homeassistant API (PLATFORMS), to see if the components of the WiLight API are allowed in this Homeassistant integration. self._components is updated by the _self_update function.
I am renaming self._components to self._wilight_components and _self_update to _wilight_update for clarity.
If you really want me to create a new constant, I will, but it doesn't make sense because I will have to update PLATFORMS and the new constant equally in an upcoming PR to include new components.

leofig-rj replied:
This conversation made me review the concept of this comparison and I saw that where any was, in reality it should be all.
In this way, only if ALL WiLight components are allowed for the HA integration version.
This occurs in the case of WiLight for ceiling fan, which has the light component and the fan on the same device.

         allowed_components = PLATFORMS
         component_ok = all (
             wilight_component in allowed_components
             for wilight_component in self._wilight_components
         )

MartinHjelmare commented on tests/components/wilight/__init__.py:
```diff
@@ -0,0 +1,76 @@
+from tests.common import MockConfigEntry
+
+HOST = "127.0.0.1"
+WILIGHT_ID = "WL000000000099"
```
```suggestion
WILIGHT_ID = "000000000099"
```

MartinHjelmare commented on tests/components/wilight/test_config_flow.py:
```diff
@@ -0,0 +1,108 @@
+    assert result["type"] == RESULT_TYPE_FORM
+    assert result["step_id"] == "confirm"
+    assert result["description_placeholders"] == {
+        CONF_NAME: WILIGHT_ID,
```
```suggestion
        CONF_NAME: f"WL{WILIGHT_ID}",
```

MartinHjelmare commented on tests/components/wilight/test_config_flow.py:
```diff
@@ -0,0 +1,108 @@
+    assert result["type"] == RESULT_TYPE_FORM
+    assert result["step_id"] == "confirm"
+    assert result["description_placeholders"] == {
+        CONF_NAME: WILIGHT_ID,
```
```suggestion
        CONF_NAME: f"WL{WILIGHT_ID}",
```

MartinHjelmare commented on tests/components/wilight/test_config_flow.py:
```diff
@@ -0,0 +1,108 @@
+    )
+
+    assert result["type"] == RESULT_TYPE_CREATE_ENTRY
+    assert result["title"] == WILIGHT_ID
```
```suggestion
    assert result["title"] == f"WL{WILIGHT_ID}"
```

MartinHjelmare commented on tests/components/wilight/test_config_flow.py:
```diff
@@ -0,0 +1,108 @@
+
+async def test_ssdp_device_exists_abort(hass: HomeAssistantType) -> None:
+    """Test we abort SSDP flow if WiLight already configured."""
+    await setup_integration(hass)
```
We don't need to set up the whole integration here. Just add a mock config entry.

leofig-rj replied:
I think I already do that, only the name is inappropriate.
I'll change the name from setup_integration to mock_config_entry

MartinHjelmare replied:
No it's setting up the integration.

leofig-rj replied:
I'm sorry but that is not clear to me.
This test is to compare a new SSDP discovery using discovery_info = MOCK_SSDP_DISCOVERY_INFO.copy (), with WiLights already configured, using mock_config_entry (hass).
Please, how could I do this without set up the whole integration?

MartinHjelmare commented on homeassistant/components/wilight/strings.json:
```diff
@@ -0,0 +1,19 @@
+    },
+    "abort": {
+      "already_in_progress": "WiLight configuration is already in progress.",
+      "already_configured": "This WiLight is already configured.",
```
Please use as many references as possible. Here they are:
https://github.com/home-assistant/core/blob/24a16ff8febff2040c00b83b42b04e02233854fa/homeassistant/strings.json#L19-L58

Here's how to reference:
https://github.com/home-assistant/core/blob/111c2006c8cf37a0d26995c0048c26780f731017/homeassistant/components/adguard/strings.json#L7

leofig-rj replied:
I reviewed the strings along with config_flow.py and removed some unnecessary ones

leofig-rj committed:
Update tests/components/wilight/__init__.py

Co-authored-by: Martin Hjelmare <marhje52@gmail.com>

leofig-rj committed:
Update tests/components/wilight/test_config_flow.py

Co-authored-by: Martin Hjelmare <marhje52@gmail.com>

leofig-rj committed:
Update tests/components/wilight/test_config_flow.py

Co-authored-by: Martin Hjelmare <marhje52@gmail.com>

leofig-rj committed:
Update tests/components/wilight/test_config_flow.py

Co-authored-by: Martin Hjelmare <marhje52@gmail.com>

leofig-rj committed:
Various updates

Various updates based on conversations with Martin

MartinHjelmare commented on tests/components/wilight/__init__.py:
```diff
@@ -0,0 +1,76 @@
+
+    entry.add_to_hass(hass)
+
+    await hass.config_entries.async_setup(entry.entry_id)
```
This sets up the integration and config entry. It's not just adding a mock config entry.

leofig-rj replied:
You're right.
In fact, this is what I need for some tests.

MartinHjelmare commented on homeassistant/components/wilight/strings.json:
```diff
@@ -0,0 +1,16 @@
+      }
+    },
+    "abort": {
+      "already_configured": "This WiLight is already configured",
```
Please use a reference as I linked before.

leofig-rj replied:
I tried to change "already_configured" to "already_configured_device" but when I ran test_config_flow.py / test_ssdp_device_exists_abort, it gave an error because "already_configured" is created automatically by _abort_if_unique_id_configured() used in config_flow..

leofig-rj committed:
Updates

link pywilight version 0.0.65
config_flow.py
testing ceiling fan WiLight as not supported

leofig-rj committed:
Update homeassistant/components/wilight/strings.json

Co-authored-by: Martin Hjelmare <marhje52@gmail.com>

leofig-rj committed:
Updates

Updates based on the conversation with Martin on August 21, 2020

leofig-rj committed:
Update to coverage

leofig-rj committed:
Update to coverage

leofig-rj committed:
Update to coverage

MartinHjelmare reviewed (approved):
Looks good!

leofig-rj commented:
Martin,

Martin,

Thank you very much for the support in this integration.
It was a great challenge for me, because it was the first time that I programmed with Python and that I participated in an Open Source project of this size. I had to learn several concepts.

Sincerely.

I have other components like fan, cover and switch. Please, can you let me know if I can make more than one component per PR?

Thank you.

Leonardo

MartinHjelmare commented:
Please add one platform per PR sequentially. Thanks! :+1:

leofig-rj commented:
👍

leofig-rj commented:
Martin,
PR https://github.com/home-assistant/home-assistant.io/pull/13547 is in Review required. Please, is there anything I need to do?

MartinHjelmare commented:
Just be patient. The docs team will see to it.

leofig-rj commented:
Thanks
